client_handlers.go

```diff
@@ -5,64 +5,82 @@ import (
 	"fmt"
 )
 
-func clientHandshakeHandler(c *Conn) (*alert, error) {
-	initalizeCipherSuite := func(h *handshakeMessageServerKeyExchange) (*alert, error) {
-		clientRandom, err := c.state.localRandom.Marshal()
+func initalizeCipherSuite(c *Conn, h *handshakeMessageServerKeyExchange) (*alert, error) {
+	clientRandom, err := c.state.localRandom.Marshal()
+	if err != nil {
+		return &alert{alertLevelFatal, alertInternalError}, err
+	}
+	serverRandom, err := c.state.remoteRandom.Marshal()
+	if err != nil {
+		return &alert{alertLevelFatal, alertInternalError}, err
+	}
+
+	if c.state.extendedMasterSecret {
+		var sessionHash []byte
+		sessionHash, err = c.handshakeCache.sessionHash(c.state.cipherSuite.hashFunc())
 		if err != nil {
 			return &alert{alertLevelFatal, alertInternalError}, err
 		}
-		serverRandom, err := c.state.remoteRandom.Marshal()
+
+		c.state.masterSecret, err = prfExtendedMasterSecret(c.state.preMasterSecret, sessionHash, c.state.cipherSuite.hashFunc())
+		if err != nil {
+			return &alert{alertLevelFatal, alertIllegalParameter}, err
+		}
+	} else {
+		c.state.masterSecret, err = prfMasterSecret(c.state.preMasterSecret, clientRandom, serverRandom, c.state.cipherSuite.hashFunc())
 		if err != nil {
 			return &alert{alertLevelFatal, alertInternalError}, err
 		}
+	}
 
-		var preMasterSecret []byte
-		if c.localPSKCallback != nil {
-			var psk []byte
-			if psk, err = c.localPSKCallback(h.identityHint); err != nil {
-				return &alert{alertLevelFatal, alertInternalError}, err
-			}
-
-			preMasterSecret = prfPSKPreMasterSecret(psk)
+	if err := c.state.cipherSuite.init(c.state.masterSecret, clientRandom, serverRandom /* isClient */, true); err != nil {
+		return &alert{alertLevelFatal, alertInternalError}, err
+	}
 
-		} else {
-			if c.localKeypair, err = generateKeypair(h.namedCurve); err != nil {
-				return &alert{alertLevelFatal, alertIllegalParameter}, err
+	if c.localPSKCallback == nil {
```

Comment:
Is it possible for `insecureSkipVerify` or `verifyPeerCertificate` to be set when dealing with PSK? Hopefully we have error handling higher up (if we don't mind ticketing!) 

I am tempted to say drop the `c.localPSKCallback` check. I like that this reads as a long chain of operations. I don't want us to assume internal state is always hostile (otherwise we might miss things sometime and trip up)

This does disagree with `defensive coding` so I am a little torn

Reply:
Ignore this! Not code you added, I will fix myself!